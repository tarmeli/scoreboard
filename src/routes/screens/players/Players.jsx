import React from "react";
import { Button, Div, Form, Input, Table } from "../../../components/Element/";
import { Tbody, Td, Tr } from "../../../components/Element/Table/";
import Map from "../../../components/utils/Map/Map";
import PlayersProvider from "./context/provider/PlayersProvider";
import PlayersHook from "./context/hook/PlayersHook";
import PlayersTableHeader from "./PlayersTableHeader/PlayersTableHeader";
import LoadingTentative from "../../../components/LoadingTentative/LoadingTentative";

const Players = ({ data: { players } }) => {
  return (
    <PlayersProvider>
      <PlayersHook>
        {({
          isLoading,
          handleDecrementWins,
          handleIncrementWins,
          handleDecrementLosses,
          handleIncrementLosses,
          handleAddNewPlayer,
          handleRemovePlayer,
          newPlayerNameInputValue,
          handleNewPlayerNameChange,
          submitNewPlayerButtonIsDisabled,
        }) => (
          <Div flexContainer flexColumn>
            <Table flexItem>
              <PlayersTableHeader />

              <Tbody>
                <Map items={players}>
                  {({ name, wins, losses, _id: id }) => {
                    return (
                      <Tr>
                        <Td>{name}</Td>

                        <Td>
                          <Button
                            onClick={async () => await handleDecrementWins(id)}
                          >
                            -
                          </Button>
                        </Td>

                        <Td>{wins}</Td>

                        <Td>
                          <Button
                            onClick={async () => await handleIncrementWins(id)}
                          >
                            +
                          </Button>
                        </Td>

                        <Td />

                        <Td>
                          <Button
                            onClick={async () =>
                              await handleDecrementLosses(id)
                            }
                          >
                            -
                          </Button>
                        </Td>

                        <Td>{losses}</Td>

                        <Td>
                          <Button
                            onClick={async () =>
                              await handleIncrementLosses(id)
                            }
                          >
                            +
                          </Button>
                        </Td>

                        <Td>
                          <Button
                            fullWidth
                            onClick={async () => await handleRemovePlayer(id)}
                          >
                            <LoadingTentative isLoading={isLoading}>
                              Remove Player
                            </LoadingTentative>
                          </Button>
                        </Td>
                      </Tr>
                    );
                  }}
                </Map>
              </Tbody>
            </Table>

            <Form
              flexItem
              fullWidth
              onSubmit={async () => await handleAddNewPlayer()}
            >
              <Div flexContainer>
                <Input
                  flexItem
                  fullWidth
                  value={newPlayerNameInputValue}
                  onChange={handleNewPlayerNameChange}
                />

                <Button
                  flexItem
                  fullWidth
                  type="submit"
                  disabled={submitNewPlayerButtonIsDisabled || isLoading}
                >
                  <LoadingTentative isLoading={isLoading}>
                    Add Player
                  </LoadingTentative>
                </Button>
              </Div>
            </Form>
          </Div>
        )}
      </PlayersHook>
    </PlayersProvider>
  );
};

export default Players;
