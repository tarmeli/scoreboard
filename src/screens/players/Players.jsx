import React from "react";

import Div from "../../components/Element/Div/Div";
import Map from "../../components/Utility/Map/Map";
import Table from "../../components/Element/Table/Table";
import Tr from "../../components/Element/Table/Tr/Tr";
import Td from "../../components/Element/Table/Td/Td";
import Th from "../../components/Element/Table/Th/Th";
import Tbody from "../../components/Element/Table/Tbody/Tbody";
import Thead from "../../components/Element/Table/Thead/Thead";
import Button from "../../components/Element/Button/Button";
import { usePlayers } from "../../data/context/players/PlayersProvider";
import LoadingTentative from "../../components/LoadingTentative/LoadingTentative";
import ErrorTentative from "../../components/ErrorTentative/ErrorTentative";

export const dependencies = {
    usePlayers,
};

const Players = () => {
    const { isLoading, players, error } = dependencies.usePlayers();

    const handleDecrementWins = (player) => {
        console.log(player);
    };

    const handleIncrementWins = (player) => {
        console.log(player);
    };

    const handleIncrementLosses = (player) => {
        console.log(player);
    };

    const handleDecrementLosses = (player) => {
        console.log(player);
    };

    return (
        <LoadingTentative isLoading={isLoading}>
            <ErrorTentative isError={error.isError} message={error.message}>
                <Div>
                    <Table>
                        <Thead>
                            <Tr>
                                <Th>Player name</Th>

                                <Th />

                                <Th>Wins</Th>

                                <Th />
                                <Th />
                                <Th />

                                <Th>Losses</Th>
                            </Tr>
                        </Thead>

                        <Tbody>
                            <Map items={players}>
                                {({ name, wins, losses }) => (
                                    <Tr>
                                        <Td>{name}</Td>

                                        <Td>
                                            <Button
                                                onClick={() =>
                                                    handleDecrementWins(name)
                                                }
                                            >
                                                -
                                            </Button>
                                        </Td>

                                        <Td>{wins}</Td>

                                        <Td>
                                            <Button
                                                onClick={() =>
                                                    handleIncrementWins(name)
                                                }
                                            >
                                                +
                                            </Button>
                                        </Td>

                                        <Td />

                                        <Td>
                                            <Button
                                                onClick={() =>
                                                    handleDecrementLosses(name)
                                                }
                                            >
                                                -
                                            </Button>
                                        </Td>

                                        <Td>{losses}</Td>

                                        <Td>
                                            <Button
                                                onClick={() =>
                                                    handleIncrementLosses(name)
                                                }
                                            >
                                                +
                                            </Button>
                                        </Td>
                                    </Tr>
                                )}
                            </Map>
                        </Tbody>
                    </Table>
                </Div>
            </ErrorTentative>
        </LoadingTentative>
    );
};

export default Players;
