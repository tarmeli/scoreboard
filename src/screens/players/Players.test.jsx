import React from "react";
import { mount } from "enzyme";

import Players, { dependencies } from "./Players";

describe("Players", () => {
    let component;
    let usePlayersMock;

    beforeEach(() => {
        usePlayersMock = jest.fn();
        dependencies.usePlayers = usePlayersMock;

        usePlayersMock.mockImplementation(() => ({
            isLoading: false,
            error: { isError: false, message: "" },
            players: [],
        }));

        component = mount(<Players />);
    });

    it("renders", () => {
        expect(component.render()).toMatchSnapshot();
    });

    it("given data is loading, when rendered, shows loading indicator", () => {
        usePlayersMock.mockImplementation(() => ({
            isLoading: true,
            error: { isError: false, message: "" },
            players: [],
        }));

        component = mount(<Players />);

        expect(component.render()).toMatchSnapshot();
    });

    it("given players are loaded, when rendered, shows players and their points", () => {
        usePlayersMock.mockImplementation(() => ({
            players: [
                {
                    name: "Some player",
                    wins: 42,
                    losses: 84,
                },
                {
                    name: "Some other player",
                    wins: 126,
                    losses: 168,
                },
            ],
            isLoading: false,
            error: {
                isError: false,
                message: "",
            },
        }));

        component = mount(<Players />);

        expect(component.render()).toMatchSnapshot();
    });

    it("given an error is provided, when rendered, shows error message", () => {
        usePlayersMock.mockImplementation(() => ({
            players: [],
            isLoading: false,
            error: {
                isError: true,
                message: "Some error message",
            },
        }));

        component = mount(<Players />);

        expect(component.render()).toMatchSnapshot();
    });
});
