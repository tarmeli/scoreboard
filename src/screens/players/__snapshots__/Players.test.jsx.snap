// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Players given an error is provided, when rendered, shows error message 1`] = `
<div>
  <div>
    Some error message
  </div>
</div>
`;

exports[`Players given data is loading, when rendered, shows loading indicator 1`] = `
<div>
  Loading...
</div>
`;

exports[`Players given players are loaded, when rendered, shows players and their points 1`] = `
<div>
  <div>
    <div>
      <table>
        <thead>
          <tr>
            <th>
              Player name
            </th>
            <th />
            <th>
              Wins
            </th>
            <th />
            <th />
            <th />
            <th>
              Losses
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Some player
            </td>
            <td>
              <button>
                -
              </button>
            </td>
            <td>
              42
            </td>
            <td>
              <button>
                +
              </button>
            </td>
            <td />
            <td>
              <button>
                -
              </button>
            </td>
            <td>
              84
            </td>
            <td>
              <button>
                +
              </button>
            </td>
          </tr>
          <tr>
            <td>
              Some other player
            </td>
            <td>
              <button>
                -
              </button>
            </td>
            <td>
              126
            </td>
            <td>
              <button>
                +
              </button>
            </td>
            <td />
            <td>
              <button>
                -
              </button>
            </td>
            <td>
              168
            </td>
            <td>
              <button>
                +
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Players renders 1`] = `
<div>
  <div>
    <div>
      <table>
        <thead>
          <tr>
            <th>
              Player name
            </th>
            <th />
            <th>
              Wins
            </th>
            <th />
            <th />
            <th />
            <th>
              Losses
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
